tephra-api/src/main/java/co/cask/tephra/Transaction.java

```diff
@@ -108,18 +157,60 @@ public long getFirstShortInProgress() {
     return firstShortInProgress;
   }
 
+  /**
+   * Returns true if the given version corresponds to a transaction that was in-progress at the time this transaction
+   * started.
+   */
   public boolean isInProgress(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0;
   }
 
+  /**
+   * Returns true if the given version is present in one of the arrays of excluded versions (in-progress and
+   * invalid transactions).
+   */
   public boolean isExcluded(long version) {
     return Arrays.binarySearch(inProgress, version) >= 0
       || Arrays.binarySearch(invalids, version) >= 0;
   }
 
+  /**
+   * Returns true if the the given version corresponds to one of the checkpoint versions in the current
+   * transaction.
+   */
+  public boolean isCheckpoint(long version) {
+    return Arrays.binarySearch(checkpointWritePointers, version) >= 0;
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   *
+   * @see #isVisible(long, boolean) to exclude the current write pointer from visible versions.  This method always
+   *      includes the current write pointer.
+   */
   public boolean isVisible(long version) {
+    return isVisible(version, true);
+  }
+
+  /**
+   * Returns whether or not the given version should be visible to the current transaction.  A version will be visible
+   * if it was successfully committed prior to the current transaction starting, or was written by the current
+   * transaction (using either the current write pointer or the write pointer from a prior checkpoint).
+   *
+   * @param version the data version to check for visibility
+   * @param excludeCurrentWritePointer whether writes from the current write pointer should be visible
+   * @return true if the version is visible, false if it should be hidden (filtered)
+   */
+  public boolean isVisible(long version, boolean excludeCurrentWritePointer) {
```

Comment:
Not sure I understand this.  The flag simply allows clients to exclude the current write pointer on a per-operation basis.  There is no need to know whether the transaction has been checkpointed or not.

From what I understand, you are suggesting that:
a) when first started, a transaction is obtained with read-your-writes behavior
b) after performing a checkpoint, read-your-write will no longer apply to the current write pointer.  To read the current write pointer, you would always have to perform a new checkpoint.

Is this correct?

I actually thought that we had worked out this discussion on the JIRA.  My plan for this change was:
a) by default, read-your-write semantics apply
b) the current write pointer can be excluded from reads by setting a flag on operations
c) used together this checkpoints, this allows clients (Phoenix) to read all prior writes in the transaction -- prior to a UPSERT ... SELECT .., perform a checkpoint, then set the exclude attribute on the Scan operation that is run.

That preserves the current read-your-writes behavior, while providing Phoenix with the tools it needs to exclude the current write pointer when necessary.